File: vertx/server/couchdb.js
// TODO only kept for reference while developing the vertx client!
// there is not really a secure way to authenticate to CouchDB from a browser client over CORS.
// Really should have just stuck with conventional wisdom but I thought I could do some serverless, CouchApp kinda thing... oh well.

/** Provides a browser client interface to a CouchDB server.
 * 
 * On Windows the CouchDB service can be started/stopped using the net command;
 *  > net start "apache couchdb"
 * It can also be controlled using the included scripts'
 *  C:/CouchDB/bin> couchdb stop
 *  C:/CouchDB/bin> couchdb restart
 * 
 * Once started the Database can be manipulated using the Fauxton webapp at;
 *  <host:port>/_utils
 * Make sure that CORS is enabled for your domain using the settings tab.
 * 
 * @param {String} config.host - the CouchDB URL
 * @param {Number} config.port - the port of the CouchDB service
 * @param {String} config.db - the name of the database being read
*/
export default (config) => {
    // TODO Couch indexes might be a better way to do this than views...
    // then again views also provide the opportunity to transform the data in situ
    // which, as we've seen, can be fairly important when instrumentation throws us curve balls the night of the mission.

    // default design docs & view for data
    const design = 'stream';
    const view = 'chronological';

    // internal Client state
    let state = {
        host : 'localhost',
        port : 5984,
        db : 'sensors', // the test database made by the simulator
        time : null // current page in the stream
    }
    Object.assign(state, config);

    // Assemble a client object for the user
    return {open,step,seek,close};

    /** Checks that the Database which corresponds to the specified data exists and has data. */
    async function open(name, password) {

        // it now requires admin privileges
        // verify the requested database exists
        let dbs = await getDatabases();
        if (!dbs.includes(state.db))
            throw( {message: `Invalid db '${state.db}', available dbs ${JSON.stringify(dbs)}`} );

        // authorize

        // // let content = JSON.stringify( {name, password} );
        // let response = await fetch( '/_session', {
        //     method: 'POST',
        //     body: {name, password},
        //     headers: {
        //         'Content-Type': 'application/json',
        //         // 'Content-Length': content.length
        //     } }
        // );
        // return response;

        // determine the start time of the data by fetching the first document
        let document = await getDocument(0);
        if (!document.rows.length)
            throw ( {message: `The ${state.db} database does not contain any events`} );
        state.time = document.rows[0].key[0]; // the first field of the compound key is a timestamp

        // // TODO determine the end time? 
        // // let end = await getDocument( start.total_rows-1 );

        return ( {host:state.host, port:state.port, db:state.db, design, view, time:state.time} );
    }

    /** retrieves the next interval of events, and advances the current time. */
    async function step( interval ) {
        let end = getNext(state.time, interval);
        let docs = await getDocumentInterval( state.time, end );
        state.time = end;
        docs.done = docs.offset < docs.total_rows
        return docs;
    }

    /** sets the current time. */
    async function seek( to ) {
        state.time = to;
        //TODO actually try to read a record and return some data?
    }

    /** This client relies on fetch calls so there are no resources to close here really */
    async function close() {}

    /** Get all the documents in the time interval. */
    async function getDocumentInterval( startkey, endkey ) {
        // assemble the URI and arguments for the specified page
        let uri = `http://${state.host}:${state.port}/${state.db}/_design/${design}/_view/${view}?include_docs=true`;
        if (startkey) uri += `&start_key=[${startkey},""]`;
        if (endkey) uri += `&end_key=[${endkey},""]`;

        // fetch the results
        let response = await fetch(uri, {mode:'cors'});

        // check for non-network errors
        if (!response.ok)
            throw(response);

        // invoke the built in JSON reader for the response body
        let page = await response.json();
        return page;
    }// TODO we still probably want ot limit the number of documents, if not the total size...

    /** Get the 'i'th document in key order from the database. */
    async function getDocument( index ) {
        // assemble the URI and arguments for the specified page
        let uri = `http://${state.host}:${state.port}/${state.db}/_design/${design}/_view/${view}`;
        let args = `?include_docs=true&limit=1&skip=${index}`

        // fetch the results
        let response = await fetch(uri+args, {mode:'cors'});

        // check for non-network errors
        if (!response.ok)
            throw(response);

        // invoke the built in JSON reader for the response body
        let page = await response.json();
        return page;
    }

    /** Calculate the time after the given interval */
    function getNext( t, dt ) {
        return t + dt;
    }// TODO we are just assuming numerical millisecond timestamps right now, but that might not always be the case...

    /** @returns An array of available databases as specified in CouchDB API. */
    async function getDatabases() {
        let response = await fetch(
            `https://${state.host}:${state.port}/_all_dbs`, 
            {}//mode:'cors'}
            );
        if (!response.ok)
            throw( response );
        let json = await response.json();
        return (json);
    }

    async function getStatus() {
        let response = await fetch(`http://${state.host}:${state.port}/`, {mode:'cors'});
        if (!response.ok)
            throw( response );
        let json = await response.json();
        return (json);
    }

    async function getSession(name, password) {
        let content = JSON.stringify( {name, password} );
        return await fetch( '/_session', {
            method: 'POST',
            body: content,
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': content.length
            } }
        );
    }

    
}

// DEBUG
//let startkey='[1562115225603,"s2"]', endkey = '[1562115229903, "r2"]';
// let startkey='[1562115225603,""]', endkey = '[1562115229903, ""]';
// let cdbclient = Couch({host:"localhost", port:5984, db:'simulation'});
// cdbclient.open()
//     .then( async(open)=>{
//         console.log(open)
//         let data = null;
//         do {
//             data = await cdbclient.step(1000);
//             console.log(data);
//         } while(data.offset < data.total_rows);
//     }).catch( (e)=>console.log(e) );
    
    // Couch's get databaseses method again, but in the chaining idiom...
    // function getDatabases() {
    //     fetch( `http://${hostname}:${port}/_all_dbs`, {mode:'cors'})
    //         //{method:'GET',mode:'cors', credentials:'include'} )
    //     .then( (response)=>{
    //         console.log(response);
    //         for (let header of response.headers)
    //             console.log(header);
    //         return response.json()
    //     })
    //     .then(json=>console.log(json))
    //     .catch( (error)=>console.log(error) );
    // }

File: vertx/server/CouchDb.java
package server;

import io.vertx.core.*;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.HttpRequest;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.predicate.ResponsePredicate;
import io.vertx.ext.web.codec.BodyCodec;

import java.util.Arrays;

/** So a Vertx client is already written for MongoDB...
 * might want to just switch to that rather than trying to write our own interface.
 * then again, will have to figure something out for interoperability with dolphin,
 * the new simulator, and we have to figure out queries and view over again...*/
public class CouchDb {

    Vertx vertx;
    WebClient client; // the client used to make HTTP requests to the CouchDB's REST API
    JsonObject token; // the current access token
    String host;
    int port;

    // TODO eventually things like current mission and roles should be stored in a user session...

    /** Obtain a session cookie using credentials in configuration.
     * https://docs.couchdb.org/en/stable/api/server/authn.html#cookie-authentication */
    public CouchDb(Vertx vertx, String host, int port) {
        this.vertx = vertx; //Vertx.currentContext().owner();
        this.client = WebClient.create(vertx);
        this.host = host;
        this.port = port;
    }

    public Future<JsonObject> getSession(String name, String password) {
        Promise<JsonObject> promise = Promise.promise();

        JsonObject credentials = new JsonObject()
                .put("name", name)
                .put("password", password);

        HttpRequest<JsonObject> session = client.post(port, host, "/_session")
                .putHeader("Content-Type", "application/json")
                .putHeader("Accept", "application/json")
                .putHeader("Content-Length", Integer.toString(credentials.toString().length()))
                .expect(ResponsePredicate.status(200, 299))
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject());

        session.sendJsonObject(credentials, request -> {

            if (!request.succeeded())
                promise.fail(request.cause());
            HttpResponse<JsonObject> response = request.result();
//            printResponse( response );

            // make sure we have admin role
            JsonObject body = response.body();
            if (body.containsKey("error"))
                promise.fail(body.getString("error"));
            if (!body.getJsonArray("roles").contains("_admin"))
                promise.fail("not an administrator");
            // TODO hmm, will I eventually need any of the role information?

            // parse and return the session cookie as a JsonObject
            String cookie = response.getHeader("Set-Cookie");
            JsonObject token = parseCookie(cookie);
            this.token = token;
            promise.complete( token );
        });
        return promise.future();
    }

    /** https://docs.couchdb.org/en/stable/api/server/common.html#all-dbs
     * @return An array of available databases as specified in CouchDB API. */
    public Future<JsonArray> getMissions() {
        Promise<JsonArray> promise = Promise.promise();

        // craft an Http request for the couch api endpoint
        String cookie = "AuthSession=" + token.getString("AuthSession");
        HttpRequest<JsonArray> dbs = client.get(port, host, "/_all_dbs")
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonArray());

        // send it asynchronously
        dbs.send( request -> {
            if (request.succeeded()) {

                HttpResponse<JsonArray> response = request.result();
                JsonArray databases = response.body();

                JsonArray missions = new JsonArray();
                databases.forEach( name -> {
                    if (!name.toString().startsWith("_"))
                        missions.add( name.toString() );
                });

//                printResponse(response);
                promise.complete(missions);
            } else
                promise.fail( request.cause() );
        });
        return promise.future();
    }

    /** Creates a database in CouchDB corresponding to a mission, and adds design documents for the needed views*/
    public Future<Void> createMission(String umi) {
        Promise<Void> promise = Promise.promise();

        String uri = "/" + umi;
        String cookie = "AuthSession=" + token.getString("AuthSession");
        HttpRequest<JsonObject> put = client.put(port, host, uri)
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject());

        put.send( request -> {
            if (request.succeeded()) {
                HttpResponse<JsonObject> response = request.result();
                JsonObject message = response.body();
                if (message.containsKey("error"))
                    promise.fail( message.toString() );
                else
                    promise.complete();
            } else
                promise.fail( request.cause() );

        });

        return promise.future();
    }

    /** Creates a database in CouchDB corresponding to a mission, and adds design documents for the needed views*/
    public Future<JsonObject> getMission( String umi ) {
        Promise<JsonObject> promise = Promise.promise();

        String uri = "/" + umi;
        String cookie = "AuthSession=" + token.getString("AuthSession");
        HttpRequest<JsonObject> mission = client.get(port, host, uri)
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject());

        mission.send( request -> {
            if (request.succeeded()) {
                HttpResponse<JsonObject> response = request.result();
                promise.complete( response.body() );
            } else
                promise.fail( request.cause() );
        });

        return promise.future();
    }

    /** Creates a database in CouchDB corresponding to a mission, and adds design documents for the needed views*/
    public Future<JsonObject> deleteMission( String umi ) {
        Promise<JsonObject> promise = Promise.promise();
        String uri = "/" + umi;

        request(HttpMethod.DELETE, uri, request-> {
            if (request.succeeded()) {
                HttpResponse<JsonObject> response = request.result();
                promise.complete( response.body() );
            } else
                promise.fail( request.cause() );
        });
        return promise.future();
    }

    private void request(HttpMethod method, String uri,
                         Handler<AsyncResult<HttpResponse<JsonObject>>> handler) {
        String cookie = "AuthSession=" + token.getString("AuthSession");
        client.request(method, port, host, uri)
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject())
                .send( handler );
    }

    public Future<JsonArray> getProducts(String umi) {
        Promise<JsonArray> promise = Promise.promise();
        String uri = "/"+umi+"/_design_docs";
        String cookie = "AuthSession=" + token.getString("AuthSession");
        HttpRequest<JsonObject> get = client.get(port, host, uri)
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject());
        get.send( request -> {
            if (request.succeeded()) {
                HttpResponse<JsonObject> response = request.result();
                JsonObject body = response.body();
                JsonArray products = new JsonArray();

                // get the list of design documents
                JsonArray rows = body.getJsonArray("rows");
                rows.forEach( row -> {

                    // trim the conventional design document prefix
                    String id = ((JsonObject)row).getString("id");
                    String view = id.substring( 1 + id.lastIndexOf("/") );

                    // the views correspond to data products
                    products.add( view );
                });
                promise.complete(products);
            } else
                promise.fail( request.cause() );
        } );
        return promise.future();
    }

    /** Adds the design document associated with the given product to the mission indicated by the umi. */
    public Future<Void> addProduct(String umi, String product) {
        Promise<Void> promise = Promise.promise();

        // get the view's map function script for the specified product
        Product.loadDesign(product).onSuccess(design -> {

            // create a request
            String uri = "/" + umi + "/_design/" + product;
            String cookie = "AuthSession=" + token.getString("AuthSession");
            int length = design.toString().length(); // redundant computation, does the request work without this header?
            HttpRequest<JsonObject> put = client.put(port, host, uri)
                    .putHeader("Accept", "application/json")
                    .putHeader("Content-Type", "application/json")
                    .putHeader("Content-Length", Integer.toString(length))
                    .putHeader("Cookie", cookie)
                    .expect(ResponsePredicate.JSON)
                    .as(BodyCodec.jsonObject());

            // send request and handle the response
            put.sendJson( design, request -> {
                if (request.succeeded()) {
                    HttpResponse<JsonObject> response = request.result();
//                    printResponse(response);
                    JsonObject message = response.body();
                    if (message.containsKey("error"))
                        promise.fail( message.toString() );
                    else
                        promise.complete();
                } else
                    promise.fail( request.cause() );
            });

        }).onFailure( promise::fail );

        return promise.future();
    }

    /** Get the 'i'th document in key order from the database. */
    public Future<JsonObject> getEvents( String umi, String product ) {
        Promise<JsonObject> promise = Promise.promise();

        // assemble the URI and arguments for the specified page
        String uri = '/' + umi
                + "/_design/" + product
                + "/_view/" + Product.DEFAULT_VIEW;
        //        "?include_docs=true&limit=1&skip=" + index; // part of a paging scheme- I don't think they do it this way anymore
        String cookie = "AuthSession=" + token.getString("AuthSession");

        // fetch the results
        HttpRequest<JsonObject> getDoc = client.get(port, host, uri)
                .putHeader("Accept", "application/json")
                .putHeader("Cookie", cookie)
                .addQueryParam("limit", "10")
//                .addQueryParam("startkey", "")
//                .addQueryParam("endkey", "")
                .expect(ResponsePredicate.JSON)
                .as(BodyCodec.jsonObject());

        getDoc.send(request -> {
            if (request.succeeded()) {
                HttpResponse<JsonObject> response = request.result();
//                printResponse(response);
                promise.complete( response.body() );
            } else
                promise.fail( request.cause() );
        });

        return promise.future();
    }
    // http://localhost:5984/sensors/_design/product/_view/events?startkey=[%222020-07-17T18:30:44.752Z%22,%22s3%22]&endkey=[%222020-07-17T18:30:44.952Z%22,%22d2%22]

    /** https://docs.couchdb.org/en/stable/api/server/authn.html#delete--_session */
    public Future<Void> deleteSession() {
        return Future.future( promise -> {

            String cookie = "AuthSession=" + token.getString("AuthSession");

            HttpRequest<JsonObject> delete = client.delete(port, host, "/_session")
                    .putHeader("Accept", "application/json")
                    .putHeader("Cookie", cookie )//token.getString("AuthSession"))
                    .expect(ResponsePredicate.status(200, 299))
                    .expect(ResponsePredicate.JSON)
                    .as(BodyCodec.jsonObject());

            delete.send( request -> {
                if (!request.succeeded())
                    promise.fail(request.cause());
                HttpResponse<JsonObject> response = request.result();

                JsonObject body = response.body();
                if (!body.getBoolean("ok"))
                    promise.fail(body.getString("error"));

                promise.complete();
            });
        });
    }

    public Future<Void> close() {
        Promise<Void> promise = Promise.promise();
        deleteSession()
                .onSuccess( r->promise.complete() )
                .onFailure( r->promise.fail(r.getCause()) );
        return promise.future();
    }

    /** The CouchDB authorization cookie is a semicolon delimited set of name value pairs. */
    private JsonObject parseCookie (String cookie) {
        String[] entries = cookie.split(";");
        JsonObject token = new JsonObject();
        Arrays.stream(entries).forEach(entry->{
            String[] pair = entry.split("=");
            if (pair.length==2)
                token.put(pair[0].trim(), pair[1].trim());
            else if (pair.length==1)
                token.put(pair[0].trim(), "");
        });
        return token;
    }

    private void printRequest(HttpRequest<?> request) {
        System.out.println("HEADERS:");
        request.headers().forEach( header->
            System.out.println( header.getKey()+'='+header.getValue()) );
    }

    private void printResponse(HttpResponse<?> response) {
        System.out.println("STATUS:"+response.statusCode()+"="+response.statusMessage());

        System.out.println("HEADERS:");
        response.headers().forEach( header->
                System.out.println( header.getKey()+'='+header.getValue()) );

        System.out.println("COOKIES:");
        response.cookies().forEach( System.out::println );

        System.out.println("BODY:");
        System.out.println( response.body() );

        System.out.println("TRAILERS:");
        response.trailers().forEach( System.out::println );
    }

// once I see some common patterns in the couch API I might want to make some custom response Predicates...
//    Function<HttpResponse<Void>, ResponsePredicateResult> methodsPredicate = resp -> {
//        String methods = resp.getHeader("Access-Control-Allow-Methods");
//        if (methods != null) {
//            if (methods.contains("POST")) {
//                return ResponsePredicateResult.success();
//            }
//        }
//        return ResponsePredicateResult.failure("Does not work");
//    };
}
